prost-build/src/code_generator.rs

```diff
@@ -106,10 +106,15 @@ impl <'a> CodeGenerator<'a> {
 
         if let Some(ref service_generator) = code_gen.config.service_generator {
             code_gen.path.push(6);
+            let mut generated = false; // We call finalize only if there's at least one service
```

Comment:
It may be nice to actually call it regardless.  I admit I don't really have a usecase in mind, but it seems strictly more flexible to me.

Reply:
If there's a compelling argument for keeping it the way it is, then I think I'd prefer skipping the boolean flag and just checking if the vector is empty, e.g.

+            if !file.service.is_empty() {
+                service_generator.finalize(&mut code_gen.buf);
             }

Reply:
There wasn't a strong reason for that, only a feeling that the service generator shouldn't be bothered with files that don't have any services.

But on the other hand, the logic is needlessly more complex, so let's get rid of it. The service generator can decide on its own not to touch such files.